pkg/lua: Make the module searchers conform to the API.

jcarrano opened the pull request:
### Contribution description

#### Bug 1

I messed up the specification for the Lua loader module.

The module searchers in the `require` package should return a string
if the module is not found, and not raise an error.

See: https://www.lua.org/manual/5.3/manual.html#pdf-package.searchers

If the searchers raised an error, then the search would stop after the first loader fails to find (which is the current- buggy- behaviour.) The result is that only the pure-lua module table get searched, and it is impossible to load a C extension module.

#### Bug 2

I violated my own specification, resulting in a crash where a module was requested but no modules were supplied by the user.

This used to work, but then some time ago I remove the definition of the weak symbols used for the tables, leaving them pointing to NULL by default.

### Testing procedure

I included a test script in the third commit.

jcarrano commented:
I labeled it as a bug because I was not respecting the API and that qualifies as a bug.

@danpetry Do you mind giving me a hand with this one?

jcarrano commented:
@danpetry Ping.

tcschmidt commented:
@jcarrano ping.

jcarrano commented:
Well, sorry for the delay. Testing this required creating some lua C-extension modules, which I was lazy to write and also did not have time to. I finally got around to doing it, and uncovered another bug in the process.

I added a test application.

leandrolanzieri commented on tests/lua_loader/main.c:
```diff
@@ -0,0 +1,103 @@
+{
+    int status;
+    status = lua_riot_do_module("test", lua_mem, LUA_MEM_SIZE,
+                                        LUAR_LOAD_BASE, NULL);
```
Please align properly.

jcarrano commented:
@leandrolanzieri 

> I noticed a change of behaviour compared to master, when trying to require a module that is not defined.

D'oh! good catch. I made a copypasta-error (it already felt like copypasta when I was writing the code, I should't have ignored that hunch).

I removed the ready-for build. I'll add it back once I have squashed.

leandrolanzieri commented:
@jcarrano that seems to fix it. Maybe you could add '\n' to the error strings, currently the output looks like:

```
Runtime error   [string "return require "something""]:1: module 'something' not found:
        no field package.preload['something']Module not found in Lua-builtinsModule not found in C-builtins
```

jcarrano commented:
I added some newline and tabs, it should be prettier now.

leandrolanzieri commented on tests/lua_loader/tests/01-run.py:
```diff
@@ -0,0 +1,36 @@
+import sys
+
+MODULE_QUERIES = [
+    ("m1", "a", "1"),
```
I just realized, if the expected value matches the module name, then the `expect_exact` will pass no matter the returned value, right?

jcarrano replied:
Dang! you are right! For some reason I thought expect_exact meant "expect exactly this". It seems I'm not be smart enough to use pexpect.

But wait, there's even more! Say I changed the 1 with a 9, then thanks to glorious pyterm the
"2019-05-31 ....!" at the beginning of a line would also be matched. It seems I'm not smart enough to use pyterm either.

leandrolanzieri commented:
@danpetry any other comments?
